aws/resource_aws_connect_instance.go

```diff
@@ -0,0 +1,284 @@
+		val := d.Get(rKey)
+		err := resourceAwsConnectInstanceUpdateAttribute(ctx, conn, d.Id(), att, strconv.FormatBool(val.(bool)))
+		if err != nil {
+			return diag.FromErr(fmt.Errorf("error setting Connect instance (%s) attribute (%s): %s", d.Id(), att, err))
+		}
+	}
+
+	return resourceAwsConnectInstanceRead(ctx, d, meta)
+}
+
+func resourceAwsConnectInstanceUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*AWSClient).connectconn
+
+	for att := range resourceConnectInstanceAttributesMapping {
+		rKey := resourceConnectInstanceAttributesMapping[att]
+		if d.HasChange(rKey) {
+			_, n := d.GetChange(rKey)
+			err := resourceAwsConnectInstanceUpdateAttribute(ctx, conn, d.Id(), att, strconv.FormatBool(n.(bool)))
+			if err != nil {
+				return diag.FromErr(fmt.Errorf("error updating Connect instance (%s) attribute (%s): %s", d.Id(), att, err))
+			}
+		}
+	}
+
+	return nil
+}
+func resourceAwsConnectInstanceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*AWSClient).connectconn
+
+	input := connect.DescribeInstanceInput{
+		InstanceId: aws.String(d.Id()),
+	}
+
+	log.Printf("[DEBUG] Reading Connect Instance %s", d.Id())
+
+	output, err := conn.DescribeInstanceWithContext(ctx, &input)
+
+	if isAWSErr(err, connect.ErrCodeResourceNotFoundException, "") {
+		log.Printf("[WARN] Connect Instance (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error reading Connect Instance (%s): %s", d.Id(), err))
+	}
+
+	instance := output.Instance
+
+	d.SetId(aws.StringValue(instance.Id))
+
+	d.Set("arn", instance.Arn)
+	d.Set("created_time", instance.CreatedTime.Format(time.RFC3339))
+	d.Set("identity_management_type", instance.IdentityManagementType)
+	d.Set("instance_alias", instance.InstanceAlias)
+	d.Set("inbound_calls_enabled", instance.InboundCallsEnabled)
+	d.Set("outbound_calls_enabled", instance.OutboundCallsEnabled)
+	d.Set("status", instance.InstanceStatus)
+	d.Set("service_role", instance.ServiceRole)
+
+	for att := range resourceConnectInstanceAttributesMapping {
+		value, err := resourceAwsConnectInstanceReadAttribute(ctx, conn, d.Id(), att)
+		if err != nil {
+			return diag.FromErr(fmt.Errorf("error reading Connect instance (%s) attribute (%s): %s", d.Id(), att, err))
+		}
+		d.Set(resourceConnectInstanceAttributesMapping[att], value)
+	}
+
+	return nil
+}
+
+func resourceAwsConnectInstanceDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*AWSClient).connectconn
+
+	input := &connect.DeleteInstanceInput{
+		InstanceId: aws.String(d.Id()),
+	}
+
+	log.Printf("[DEBUG] Deleting Connect Instance %s", d.Id())
+
+	_, err := conn.DeleteInstance(input)
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error deleting Connect Instance (%s): %s", d.Id(), err))
+	}
+
+	// For using an existing directory,
```

Comment:
In this case a wait should be added in the directory resource. the sleep 30 seems a bit arbitrary.

Reply:
Yeah, the wait is definitely arbitrary because there is no API call that I found that can check if the Connect Instance was deregistered from the Directory Service or not. Happy to change this if you have a better idea, Ideally I want to make sure the instance was dissociated from the directory and I was running into this issue while when the test ran because it will immediately.

Reply:
See other resources delete logic where they wait for a certain result no skip a predefined error until it doesn’t happen or retry until an error doesn’t appear anymore (and this case it means the AD is not connected to the instance)